[whats-new] Expand 13.3 release notes (#838)
microsoft/aspire.dev: src/frontend/src/content/docs/whats-new/aspire-13-3.mdx

```bash title="Aspire CLI — Run the standalone dashboard"
aspire dashboard run
```

```text frame="terminal" title="Output"
   Dashboard:  http://localhost:18888/login?t=a360442eeb99c38fe60954dc4f045acc

   OTLP/gRPC:  http://localhost:4317

   OTLP/HTTP:  http://localhost:4318

        Logs:  ~/.aspire/logs/cli_20260506T065505_02e43b93.log
```

The command is interactive and blocking — it stays in the foreground while the dashboard is running. Open the **Dashboard** URL in your browser, then point any OTLP-compatible application at the **OTLP/gRPC** or **OTLP/HTTP** endpoint, and your logs, traces, and metrics show up live. The container image is still available for environments where running the CLI isn't an option.

<LearnMore>
  Command reference: [`aspire dashboard`](/reference/cli/commands/aspire-dashboard/) and [Standalone dashboard](/dashboard/standalone/).
</LearnMore>

### `aspire docs api` — API reference search from the CLI

Search and read the Aspire API reference directly from the terminal:

```bash title="Aspire CLI — Search and view API docs"
aspire docs api list
aspire docs api search "AddAzureFrontDoor"

> aspire docs api get Aspire.Hosting.AzureFrontDoorBuilderExtensions

Reviewer: ```suggestion
aspire docs api list typescript
aspire docs api search "WithReference"
aspire docs api get typescript/aspire.hosting/withreference
```

`aspire ps`, `aspire describe`, and other CLI commands now hide resources that are marked as hidden in the AppHost (such as proxies, helper containers, and migrations). Use `--include-hidden` to include them.

### `aspire do` improvements

- **`--list-steps`** prints the list of pipeline steps that would run for `aspire do`, `aspire publish`, `aspire deploy`, or `aspire destroy` without actually executing them.
- **`check-container-runtime`** is a built-in pipeline step that fails fast when no container runtime is available, preventing late-stage build/publish failures.
- **Publish summary** now shows a hierarchical, timeline-style view of pipeline execution.
- **Independent steps continue on sibling failure** — a single failed step no longer blocks unrelated work in the same pipeline run.

### CLI quality-of-life

- **AppHost path guardrails** — the CLI's global config now validates AppHost paths to prevent accidentally pointing at the wrong project.
- **Container runtime health check** runs before `aspire deploy` to catch missing/broken Docker or Podman setups early.
- **`aspire ps`** displays the dashboard URL alongside running AppHosts.

> - **Non-interactive CLI mode** has been improved with the new `PromptBinding` that surfaces required values up front.

Reviewer 2: ```suggestion
- **Non-interactive CLI mode** has been improved. Many CLI commands have new options to support using them with `--non-interactive`.
```

The notification center is accessible via the bell icon in the top-right corner of the header. When there are unread notifications, a badge with the unread count appears on the icon.

<Image
  src={notificationCenterBell}
  alt="The Aspire dashboard showing a notification bell icon with a badge in the top-right corner"
/>

Selecting the bell icon opens the **Notifications** dialog, which lists all notifications. From here you can view command responses.

<Image
  src={notificationCenterOpen}
  alt="The Aspire dashboard notification panel open, showing a 'Rebuild succeeded' notification with a View response button"
/>

### Telemetry API enabled by default

The dashboard's telemetry API is now **enabled by default**, with an automatic token exchange between the CLI/extension and the dashboard. AI tools and other consumers can subscribe to live telemetry without any extra configuration.

<Aside type="note">
You can disable the API via dashboard configuration if you don't want it exposed.
</Aside>

> 

Reviewer 2: Not worth calling out publically
```suggestion
```

### AI features disabled by default

The dashboard's AI features are **disabled by default** in 13.3 (`Dashboard:AI:Disabled` defaults to `true`). The dashboard's AI surface is being phased out in favor of the Aspire CLI and the `aspire` agent skill, which provide a richer terminal-based and editor-integrated experience. Future releases will remove these dashboard features entirely.

### Open dashboard in VS Code Simple Browser

When you launch your Aspire app from VS Code, the dashboard now opens in **VS Code's Simple Browser** by default rather than spawning an external browser window. The behavior is controlled by the `aspire.dashboardBrowser` setting.

> ## 🧰 VS Code extension

Reviewer 2: @adamint Improve this area. Choose a couple of big features and add sections + screenshots for them.

- **AppHost CodeLens and gutter decorations** — the extension renders inline run/stop affordances and resource state right in the editor.
- **Auto-restore** — set `aspire.enableAutoRestore` to have the extension automatically run `aspire restore` when a workspace opens or its config changes.
- **Launch config** entries can now specify `env` and `args` for the AppHost, just like a regular .NET launch config.
- **Open AppHost source** from the *Running AppHosts* view.
- **Open Aspire Dashboard** from the VS Code Command Palette.
- **Right-click context menus** on resource endpoint URLs (open, copy, etc.).
- The walkthrough has been updated to cover TypeScript AppHosts.
- The extension auto-detects the Aspire CLI in default install paths even when it isn't on `PATH`.

<LearnMore>
  For more details, see [Aspire VS Code extension](/get-started/aspire-vscode-extension/).
</LearnMore>

## 📦 Integration updates

### 🔍 Browser logs and screenshots

> The new **Aspire.Hosting.Browsers** integration captures browser console logs, network requests, and screenshots from your frontend resources during development, surfacing them in the dashboard alongside server-side telemetry. Add `WithBrowserLogs()` to any frontend resource to enable it.

Reviewer 2: A screenshot of seeing logs from the browser in the dashboard would help make this pop

This release introduces:

- **Azure Front Door** and **Network Security Perimeter** hosting integrations.
- **Azure Kubernetes Service (AKS)** hosting integration (preview).
- **Kubernetes Ingress and Gateway API routing** resources (preview), plus first-class Helm-based Kubernetes deployment.
- The **container tunnel** is now enabled by default, providing uniform host connectivity across Docker Desktop, Docker Engine, and Podman.
- A new **`aspire destroy`** command and the **`aspire dashboard`** CLI, plus the Aspire CLI is now available as a `dotnet tool`.
- Major **polyglot AppHost** parity work for TypeScript, Python, and Java (experimental).
- **`aspire init`** (preview) — bootstrap an AppHost into an existing repository with the new `aspireify` agent skill.
- **JavaScript publish methods** (preview) — a unified set of `PublishAs*` APIs for Next.js, Vite, Bun, and other JS frameworks.
- **`WithBrowserLogs()`** — capture browser console logs, network requests, and screenshots from your frontend resources.
- **Foundry Prompt Agent** support.

> - …and much more.

Reviewer 2: Add command returns and the notification center. Those are big, visible features in the dashboard.

</Aside>

<Aside type="note">
If you are on a version of the Aspire CLI less than 13, please use the
[installation instructions](/get-started/install-cli/) to install the latest
stable version.
</Aside>

For general purpose upgrade guidance, see [Upgrade Aspire](/whats-new/upgrade-aspire/).

The easiest way to upgrade to Aspire 13.3 is using the [`aspire update` command](/reference/cli/commands/aspire-update/):

<Steps>

1. Update the Aspire CLI itself:

   ```bash title="Aspire CLI — Update the CLI"
   aspire update --self
   ```

1. Update your projects (run from the root of your repository):

   ```bash title="Aspire CLI — Update all Aspire packages"
   aspire update
   ```

</Steps>

Or install the CLI from scratch:

<OsAwareTabs>
  <Fragment slot="unix">

    ```bash title="Aspire CLI — Install Aspire CLI"
    curl -sSL https://aspire.dev/install.sh | bash
    ```

  </Fragment>
  <Fragment slot="windows">

    ```powershell title="Aspire CLI — Install Aspire CLI"
    irm https://aspire.dev/install.ps1 | iex
    ```

  </Fragment>
</OsAwareTabs>

<LearnMore>
  For more details on installing the Aspire CLI, see [Install the
  CLI](/get-started/install-cli/).
</LearnMore>

> ## ☁️ Azure improvements

Reviewer 2: Do we want to lead with Azure?

From a perception standpoint, I think core features to app host, CLI, dashboard would be better to start with.

Reviewer: I agree, I think we should push this down lower.

Aspire 13.3 adds a new hosting integration for [Azure Front Door](https://learn.microsoft.com/azure/frontdoor/), Microsoft's global edge network for fast, secure, and highly available web apps. Add a Front Door profile in front of your backends with `AddAzureFrontDoor` and attach origins with `WithOrigin`:

<Tabs syncKey='aspire-lang'>

> <TabItem id='csharp' label='C# AppHost'>

Reviewer: ```suggestion
<TabItem id='csharp' label='AppHost.cs'>
```

```csharp title="C# — AppHost.cs"
// Two backend projects exposed as external HTTP endpoints.
var api = builder.AddProject<Projects.Api>("api")
    .WithExternalHttpEndpoints();

var web = builder.AddProject<Projects.Web>("web")
    .WithExternalHttpEndpoints();

// Provision an Azure Front Door profile and attach each backend as an origin.
// Each WithOrigin call creates its own endpoint, origin group, origin, and route.
builder.AddAzureFrontDoor("frontdoor")
    .WithOrigin(api)
    .WithOrigin(web);
```

</TabItem>

> <TabItem id='typescript' label='TypeScript AppHost'>

Reviewer: ```suggestion
<TabItem id='typescript' label='apphost.ts'>
```

<Tabs>
<TabItem label="C# AppHost">

```csharp title="C# — Exclude an endpoint from WithReference"
// Mark the "admin" endpoint so it isn't injected into consumers via WithReference().
builder.AddProject<Projects.Api>("api")
    .WithEndpoint("admin", e => e.ExcludeReferenceEndpoint = true);
```

</TabItem>
<TabItem label="TypeScript AppHost">

```typescript title="TypeScript — Exclude an endpoint from withReference"
import { createBuilder } from './.modules/aspire.js';

const builder = await createBuilder();

// Mark the "admin" endpoint so it isn't injected into consumers via withReference().
const api = await builder.addProject('api', '../Api/Api.csproj');
await api.withEndpoint('admin', { excludeReferenceEndpoint: true });

await builder.build().run();
```

</TabItem>
</Tabs>

### `ASPIRE_ENVIRONMENT` variable

The AppHost now honors a top-level `ASPIRE_ENVIRONMENT` environment variable that selects an environment-specific configuration profile, complementing `ASPNETCORE_ENVIRONMENT`/`DOTNET_ENVIRONMENT`.

> ### DCP uses the ASP.NET Core developer certificate

Reviewer 3: ```suggestion
### DCP use of the Aspire trusted developer certificate
```

> DCP now uses the existing ASP.NET Core developer certificate when available, improving trust on Windows, macOS, and Linux without requiring you to provision a separate Aspire-specific certificate. The behavior is controlled by `ASPIRE_DCP_DEVELOPER_CERTIFICATE`.

Reviewer 3: ```suggestion
The Microsoft Developer Control Plane (DCP) server can now use the existing Aspire trusted developer certificate on Windows if you are experiencing issues with the ephemeral certificate that DCP uses by default. This behavior is controlled by the `ASPIRE_DCP_DEVELOPER_CERTIFICATE` environment variable.
```

    });
```

</TabItem>
<TabItem id='typescript' label='TypeScript AppHost'>

```typescript title="TypeScript — Privileged Docker Compose service"
import { createBuilder } from './.modules/aspire.js';

const builder = await createBuilder();

// Add a netshoot diagnostics container and publish it to Docker Compose
// with privileged mode enabled (required for low-level network tooling).
const netshoot = await builder.addContainer('netshoot', 'nicolaka/netshoot');
await netshoot.publishAsDockerComposeService(async (resource, service) => {
    await service.privileged.set(true);
});

await builder.build().run();
```

</TabItem>
</Tabs>

### Dockerfile builder APIs for polyglot AppHosts

Polyglot AppHosts (TypeScript, Python, Java) can now build Dockerfiles programmatically with `WithDockerfileBuilder` / `AddDockerfileBuilder`, mirroring the C# `WithDockerfile` extension. Diagnostic [`ASPIREDOCKERFILEBUILDER001`](/diagnostics/aspiredockerfilebuilder001/) covers the experimental warning.

<LearnMore>
  For more details, see [Customize Dockerfile builds](/app-host/withdockerfile/).
</LearnMore>

## 🛠️ CLI enhancements

### Aspire CLI as a NativeAOT dotnet tool

> Aspire 13.3 publishes the Aspire CLI as a **NativeAOT .NET global tool**. The CLI has always shipped as a `dotnet tool`, but in 13.3 it takes advantage of new .NET 10 support for distributing NativeAOT-compiled tools — meaning instant startup with no JIT or runtime warmup, and no managed runtime dependency in the tool package. If you already have .NET 10 installed, you can install the CLI without using the curl/PowerShell installer:

Reviewer 2: Reword to put benefit first. "The Aspire CLI is faster than ever using Native AOT. Expect to see instant startup, etc..." then talk about details

Reviewer 2: Probably applies in other places. Will use AI follow up review to fine places that could use increased excitement 😄

- **Non-interactive CLI mode** has been improved with the new `PromptBinding` that surfaces required values up front.

### Improved AI agent init

`aspire agent init` gained a **location selection** step and now supports the standard Aspire skill family for one-shot scaffolding.

### Java polyglot AppHost (experimental) — welcome to the family, Java developers

Aspire 13.3 introduces **experimental** support for Java AppHosts, joining TypeScript and Python in the polyglot family. If your team lives in IntelliJ and thinks in Spring Boot, you can now orchestrate your distributed app without switching to a C# project. The Aspire CLI scaffolds, builds, and runs Java AppHosts that consume a generated Java Aspire SDK (the same `.modules`-style code-generation model used for the TypeScript AppHost). Java 17+ is required.

The feature is gated behind a feature flag — enable it with:

```bash title="Aspire CLI"
aspire config set -g features.experimentalPolyglot:java true
aspire new # then pick "Empty (Java AppHost)"
```

<Aside type="note">
The Java polyglot AppHost is experimental in 13.3 and APIs may change in future releases.
</Aside>

> ## 📊 Dashboard improvements

Reviewer 2: Is the rebuild command new? That's a useful feature and could be called out here.

The dashboard includes a notification center accessible via the bell icon in the top-right corner of the header. When there are unread notifications, a badge with the unread count appears on the icon.

<Image
  src={notificationCenterBell}
  alt="The Aspire dashboard showing a notification bell icon with a badge in the top-right corner"
/>

Selecting the bell icon opens the **Notifications** dialog, which lists all notifications. From here you can view command responses.

<Image
  src={notificationCenterOpen}
  alt="The Aspire dashboard notification panel open, showing a 'Rebuild succeeded' notification with a View response button"
/>

### Telemetry API enabled by default

The dashboard's telemetry API is now **enabled by default**, with an automatic token exchange between the CLI/extension and the dashboard. AI tools and other consumers can subscribe to live telemetry without any extra configuration.

<Aside type="note">
You can disable the API via dashboard configuration if you don't want it exposed.
</Aside>

### AI features disabled by default

> The dashboard's AI features are now **disabled by default** (`Dashboard:AI:Disabled` defaults to `true`). Set the value to `false` to opt back in.

Reviewer 2: This needs more content to explain why it's off by default.

There is an agentic coding + dashboard page in 13.3 that has details that can be placed here.

### Other dashboard polish

- **Resource state details** now appear in resource details.
- **Parameters page** refinements make secret values clearer to manage.
- **Short trace IDs** (8 hex characters) are accepted by the telemetry API for cross-tool linking.
- Resource names and lifecycle notifications are now stable across reloads.

### Open dashboard in VS Code Simple Browser

> When you launch your Aspire app from VS Code, the dashboard now opens in **VS Code's Simple Browser** by default rather than spawning an external browser window. The behavior is controlled by the `aspire.dashboardBrowser` setting.

Reviewer 2: This should be under `VS Code extensions`

<LearnMore>
  For details on the `aspire new` command and all available templates, see [aspire new command](/reference/cli/commands/aspire-new/).
</LearnMore>

## 🐛 Bug fixes and full changelog

For the complete list of bug fixes and smaller changes in this release, see the [Aspire 13.3 release notes on GitHub](https://github.com/microsoft/aspire/releases/tag/v13.3.0).

## 🙏 Community contributions

Aspire is built in the open, and this release wouldn't be what it is without you. A huge thank you to all community contributors who helped make Aspire 13.3 possible — including [@afscrome](https://github.com/afscrome) for the lifecycle Subscribe extension methods, [@spboyer](https://github.com/spboyer) for documentation and template polish, [@holystix04](https://github.com/holystix04), and [@tranhoangtu-it](https://github.com/tranhoangtu-it). We're always excited to [see community contributions](/community/contributors/)! If you'd like to get involved, check out our [contributing guide](/community/contributor-guide/).

## ⚠️ Breaking changes

> 

Reviewer: This section is missing the breaking changes to TS AppHost APIs. There are likely many more than I have listed here:

| Old method (deprecated) | Replacement |
|---|---|
| `withEnvironmentExpression(name, expr)` | `withEnvironment(name, expr)` |
| `withEnvironmentEndpoint(name, endpoint)` | `withEnvironment(name, endpoint)` |
| `withEnvironmentParameter(name, param)` | `withEnvironment(name, param)` |
| `withEnvironmentConnectionString(name, resource)` | `withEnvironment(name, resource)` |
| `withEnvironmentFromOutput(name, output)` | `withEnvironment(name, output)` |
| `withEnvironmentFromKeyVaultSecret(name, secret)` | `withEnvironment(name, secret)` |
